src/runtime/internal/database.server.ts

```diff
@@ -113,20 +115,64 @@ async function _checkAndImportDatabaseIntegrity(event: H3Event, collection: stri
 
     // Delete old version -- checksum exists but does not match with bundled checksum
     await db.exec(`DELETE FROM ${tables.info} WHERE id = ?`, [`checksum_${collection}`])
+
+    if (!unchangedStructure) {
+      // we need to drop the table and recreate it
+      await db.exec(`DROP TABLE IF EXISTS ${tables[collection]}`)
+    }
   }
 
   const dump = await loadDatabaseDump(event, collection).then(decompressSQLDump)
 
-  await dump.reduce(async (prev: Promise<void>, sql: string) => {
+  const indexesToInsert: number[] = []
+  if (unchangedStructure) {
+    // get the list of hash to insert
+    const hashListFromTheDump: string[] = JSON.parse(dump[0].slice(3))
+
+    // get the list of hash in the database
+    const hashesInDbRecords = await db.all<{ __hash__: string }>(`SELECT __hash__ FROM ${tables[collection]}`).catch(() => [] as { __hash__: string }[])
+    const hashesInDb = hashesInDbRecords.map(r => r.__hash__)
+
+    // get the list of hash to delete
+    const hashesToDelete = hashesInDb.filter(hash => !hashListFromTheDump.includes(hash))
+    if (hashesToDelete.length) {
+      await db.exec(`DELETE FROM ${tables[collection]} WHERE __hash__ IN (${hashesToDelete.map(() => '?').join(',')})`, hashesToDelete)
+    }
+
+    // get the list indexes of the queries we will insert/update
+    let listIndex = 0
+    for (const hash of hashListFromTheDump) {
+      const index = hashesInDb.indexOf(hash)
+      // if the hash was not found in db, we will insert it
+      if (index === -1) {
+        indexesToInsert.push(listIndex)
+      }
+      listIndex++
+    }
+  }
+
+  await dump.reduce(async (prev: Promise<void>, sql: string, index: number) => {
     await prev
+
+    // skip sql comment
+    if (sql.startsWith('-- ')) {
+      return Promise.resolve()
+    }
+
+    // If the structure has not changed,
+    // skip any insert/update line whose hash is already in the database.
+    // If not, since we dropped the table, no record is skipped, insert them all again.
+    if (unchangedStructure && (sql.startsWith('INSERT ') || sql.startsWith('UPDATE ')) && !indexesToInsert.includes(index)) {
```

Comment:
Values of `indexesToInsert` does not match the index of queries in `dump` therefore `indexesToInsert.includes(index)`  is not a valid condition to check

Reply:
The hashes indexed here come directly from the `dump`
1. So they are ordered the same way as the insert query, I make sure of it at build time.
2. The table of hash is padded at the beginning with the exact number of lines in the init. If there are 3 lines of insert, we insert 3 empty strings at the beginning.
3. That is why I would have loved a few e2e tests.
4. I like your idea of having a hash at the begining of the line so the code is simpler to understand